src/event/sdam/mod.rs

```diff
@@ -0,0 +1,210 @@
+}
+
+/// Published when a server monitor's `hello` or legacy hello command succeeds.
+#[derive(Clone, Debug)]
+#[non_exhaustive]
+pub struct ServerHeartbeatSucceededEvent {
+    /// The execution time of the event.
+    pub duration: Duration,
+
+    /// The reply to the `hello` or legacy hello command.
+    pub reply: Document,
+
+    /// The address of the server.
+    pub server_address: ServerAddress,
+    // TODO RUST-560 add awaited field
+}
+
+/// Published when a server monitor's `hello` or legacy hello command fails.
+#[derive(Clone, Debug)]
+#[non_exhaustive]
+pub struct ServerHeartbeatFailedEvent {
+    /// The execution time of the event.
+    pub duration: Duration,
+
+    /// The failure that occurred.
+    pub failure: Error,
+
+    /// The address of the server.
+    pub server_address: ServerAddress,
+    // TODO RUST-560 add awaited field
+}
+
+/// Applications can implement this trait to specify custom logic to run on each SDAM event sent
+/// by the driver.
+///
+/// ```rust
+/// # use std::sync::Arc;
+/// #
+/// # use mongodb::{
+/// #     error::Result,
+/// #     event::sdam::{
+/// #         SdamEventHandler,
+/// #         ServerHeartbeatFailedEvent,
+/// #     },
+/// #     options::ClientOptions,
+/// # };
+/// # #[cfg(feature = "sync")]
+/// # use mongodb::sync::Client;
+/// # #[cfg(not(feature = "sync"))]
+/// # use mongodb::Client;
+/// #
+/// struct FailedHeartbeatLogger;
+///
+/// impl SdamEventHandler for FailedHeartbeatLogger {
+///     fn handle_server_heartbeat_failed_event(&self, event: ServerHeartbeatFailedEvent) {
+///         eprintln!("Failed server heartbeat: {:?}", event);
+///     }
+/// }
+///
+/// # fn do_stuff() -> Result<()> {
+/// let handler: Arc<dyn SdamEventHandler> = Arc::new(FailedHeartbeatLogger);
+/// let options = ClientOptions::builder()
+///                   .sdam_event_handler(handler)
+///                   .build();
+/// let client = Client::with_options(options)?;
+///
+/// // Do things with the client, and failed server heartbeats will be logged to stderr.
+/// # Ok(())
+/// # }
+/// ```
+pub trait SdamEventHandler: Send + Sync {
+    /// A [`Client`](../../struct.Client.html) will call this method on each registered handler when
+    /// a server description changes.
+    fn handle_server_description_changed_event(&self, _event: ServerDescriptionChangedEvent) {}
+
+    /// A [`Client`](../../struct.Client.html) will call this method on each registered handler when
+    /// a server is initialized.
+    fn handle_server_opening_event(&self, _event: ServerOpeningEvent) {}
+
+    /// A [`Client`](../../struct.Client.html) will call this method on each registered handler when
+    /// a server is closed.
+    fn handle_server_closed_event(&self, _event: ServerClosedEvent) {}
+
+    /// A [`Client`](../../struct.Client.html) will call this method on each registered handler when
+    /// its topology description changes. Note that this method will not be called until the
```

Comment:
Was this note about not being called until the `Client` is dropped meant to be included with the topolgy closed handler method?

Reply:
yep good catch, fixed